feat(button): handle style icons as children inside of buttons

akornmeier opened the pull request:
- **add**: startIcon and endIcon props that will place an `Icon` in `Button` either before or after button text
- **add**: tests to verify proper elements are added when an Icon is passed to either prop
- **refactor**: update `Button` code to be a Stateless Functional Component

Closes: #141

scurker commented on packages/react/src/components/Button/index.tsx:
```diff
@@ -1,42 +1,40 @@
       className,
-      buttonRef,
-      thin,
+      startIcon,
```
I think we should favor composition over `propTypes` here. The feels way too prescriptive. The icon should naturally be placed in the button where it's needed:

```jsx
<Button>
  <Icon type="icon-type">
  Submit
</Button>
```

or

```jsx
<Button>
  Submit
  <Icon type="icon-type">
</Button>
```

It would be better to have inheritance styles [here](https://github.com/dequelabs/cauldron/blob/develop/packages/styles/button.css) so the icon would naturally inherit styles from the button. You're already on the right track with setting `flex` on the button, just need the appropriate styles for the icon.

```css
.Button--primary .Icon,
.Button--secondary .Icon,
.Button--clear .Icon,
.Button--error .Icon {
  /* icon styles */
}
```

scurker commented on packages/styles/button.css:
```diff
@@ -80,6 +83,24 @@ button.Link {
+
+.Button--thin .Button--start-icon svg,
+.Button--thin .Button--end-icon svg {
+  width: 15px;
```
This should be calculated from the button height. Something like:

```css
width: calc(var(--button-thin-height) - 8px);
height: calc(var(--button-thin-height) - 8px);
```

scurker commented on packages/styles/button.css:
```diff
@@ -80,6 +83,24 @@ button.Link {
 
+.Button--start-icon {
+  display: inherit;
+  margin-right: 4px;
```
What are the margins used for? Is there any way we can avoid this?

scurker commented on packages/react/src/components/Button/index.tsx:
```diff
@@ -1,42 +1,40 @@
-  render() {
-    const {
-      variant,
+const Button = forwardRef<HTMLButtonElement, ButtonProps>(
```
I'm in favor of using `forwardRef` here, but this is a breaking change. I think we should eventually change this, but I don't think it should be a part of this PR.

akornmeier commented:
Update: 
- Removed Icon related props in favor of composition
- Added support for both buttonRef and forwardRef in component (buttonRef will trigger a console.warn deprecated message)
- Updated css to handle the positioning/alignment of the <Icon> element passed in as children
- updated tests

scurker commented on packages/react/src/components/Button/index.tsx:
```diff
@@ -1,42 +1,31 @@
+    }: ButtonProps,
+    ref
+  ) => {
+    if (buttonRef) {
+      console.warn(
+        "%c Warning: 'buttonRef' prop is deprecated, please use 'ref'. ",
+        'background: #222; color: #bada44'
+      );
+    }
```
Let's leave out the `console.warn` here.

akornmeier replied:
removed in the latest commit

scurker reviewed (changes requested):
One small nitpick
